### PR TITLE
Add possibility to define bidirectional migrations

### PR DESCRIPTION
**Pull request by psliwa**
This PR is for #57.

DSL hasn't changed, api is simple. If you want to define backward migration, you can use `backTo` function on `Migrator`. I've assumed not for every forward migration has to exist backward migration (so `backTo` invocation is optional), because when forward migration eg. adds new field, backward migration is not needed. The second reason for optional `backTo` migration is just not using backward migrations in the project.

Usage from tests:

```scala
val migrator = from[String, V1]
   .to[V2](_ + "V2")
   .backTo[V1](_.replace("V2", ""))
   .to[V3](_ + "V3")
   .backTo[V2](_.replace("V3", ""))
   .to[V4](_ + "V4")
   .backTo[V3](_.replace("V4", ""))
   .to[V5](_ + "V5")
```

### CODE REVIEW

File: stamina-core/src/test/scala/stamina/migrations/MigratorSpec.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package stamina.migrations
+
+import stamina._
+
+class MigratorSpec extends StaminaSpec {
+
+  val migrator = from[String, V1]
```
Comment (psliwa) [inline stamina-core/src/test/scala/stamina/migrations/MigratorSpec.scala]: This test suits to scalacheck, do you think I can add scalacheck as dependency and rewrite this test?

File: stamina-core/src/test/scala/stamina/migrations/MigratorSpec.scala [hunk 2]
```diff
@@ -0,0 +1,85 @@
+package stamina.migrations
+
+import stamina._
+
+class MigratorSpec extends StaminaSpec {
```
Comment (psliwa) [inline stamina-core/src/test/scala/stamina/migrations/MigratorSpec.scala:5]: This test suits to scalacheck, do you think I can add scalacheck as dependency and rewrite this test?

**Comment by psliwa** [comment]
@agemooij Any chance to take a look at this PR soon?

**Comment by agemooij** [comment]
@raboof what do you think of this PR?

File: stamina-core/src/main/scala/stamina/migrations.scala [hunk 3]
```diff
@@ -12,21 +12,11 @@ package object migrations {
...
 
   /**
-   * Adds support for combining two instances of Migration[T] into a new
-   * Migration[T] that will apply the first one and then the second one.
-   */
-  implicit class MigrationWithComposition[T](val firstMigration: Migration[T]) extends AnyVal {
-    def &&(secondMigration: Migration[T]): Migration[T] = {
-      (value: T) ⇒ secondMigration(firstMigration(value))
-    }
-  }
```
Comment (agemooij) [inline stamina-core/src/main/scala/stamina/migrations.scala:22]: Why was this removed?

File: stamina-core/src/main/scala/stamina/migrations.scala [hunk 4]
```diff
@@ -12,21 +12,11 @@ package object migrations {
...
-
-  /**
    * Creates a Migrator[T, V1] that can function as a builder for
    * creating Migrator[T, V2], etc. Its migration will be the identity
    * function so calling its migrate function will not have any effect.
    */
-  def from[T, V <: V1: VersionInfo]: Migrator[T, V] = new Migrator[T, V](Map(Version.numberFor[V] → identityMigration[T]))
+  def from[T, V <: V1: VersionInfo]: Migrator[T, V] = new Migrator[T, V](Map(Version.numberFor[V] → identityMigration[T]), Map.empty)
```
Comment (agemooij) [inline stamina-core/src/main/scala/stamina/migrations.scala:19]: Why not make the empty map a default argument of the `Migrator` class?

**Review by agemooij — COMMENTED** [review]
This code seems to always apply all forward and backward migrations instead of allowing them to cancel each other out. Have you considered removing migrations if they would be cancelled out by a backward one?

File: stamina-core/src/main/scala/stamina/migrations.scala (hunk 3 above)
Comment (psliwa) [reply stamina-core/src/main/scala/stamina/migrations.scala:22]: Because it is not used anymore.

File: stamina-core/src/main/scala/stamina/migrations.scala (hunk 4 above)
Comment (psliwa) [reply stamina-core/src/main/scala/stamina/migrations.scala:19]: Because backward migrations are empty when we start defining migration - so empty map is natural initial value.

**Comment by psliwa** [comment]
`This code seems to always apply all forward and backward migrations instead of allowing them to cancel each other out. Have you considered removing migrations if they would be cancelled out by a backward one?`

I am not sure what you mean under `always apply all forward and backward migrations instead of allowing them to cancel each other out`. The idea is to have migrations decoupled from persisters, so I can use migrations alone. When I want to migrate something from version X to Y, `Migrator` should apply transformations. If those transformations are forward or backward - it depends on given versions.

`Have you considered removing migrations if they would be cancelled out by a backward one?`
I don't fully understand, there is no sense in removing migrations when backward one is added. They coexists. Which one is used (backward or forward) depends on if `fromVersion` is greater or lower than `toVersion`. In Persisters only forward migrations are used, backward migrations we would like to use in our rollback tool.

**Comment by psliwa** [comment]
@agemooij ping

**Comment by psliwa** [comment]
Any progress? Maybe something is unclear?

**Comment by psliwa** [comment]
@agemooij @raboof ? :pray:

**Comment by agemooij** [comment]
@psliwa Sorry man. Stamina has been so completely out of scope for me lately that it's hard to spend time on it .

I had some minor review comments before that you haven't answered yet but overall I'm ok with merging this.

@raboof? What do you think?

File: stamina-core/src/main/scala/stamina/migrations.scala [hunk 5]
```diff
@@ -55,20 +45,36 @@ package migrations {
    * }}}
    *
    *  @tparam R The type of raw data being migrated. In the JSON implementation this would be `JsValue`.
-   *  @tparam V The "current" version of this Migrator, i.e. it can migrate values from V1 to this version or any version in between.
+   *  @tparam V The "current" version of this Migrator.
    */
-  class Migrator[R, V <: Version: VersionInfo] private[stamina] (migrations: Map[Int, Migration[R]] = Map.empty) {
-    def canMigrate(fromVersion: Int): Boolean = migrations.contains(fromVersion)
+  class Migrator[R, V <: Version: VersionInfo] private[stamina] (forwardMigrations: Map[Int, Migration[R]] = Map.empty, backwardMigrations: Map[Int, Migration[R]] = Map.empty) {
+    def canMigrate(fromVersion: Int, toVersion: Int): Boolean = forwardMigrations.contains(fromVersion) && forwardMigrations.contains(toVersion)
```
Comment (raboof) [inline stamina-core/src/main/scala/stamina/migrations.scala:51]: Is this correct? It's not obvious to me that we now also need `forwardMigrations.contains(toVersion)`, and shouldn't this also take into account that we can support backwards migrations?

**Review by raboof — COMMENTED** [review]
@agemooij @psliwa sorry for taking so long to respond too.

This looks reasonable, might be good to address the couple of small comments given on the PR, and perhaps add some documentation? But otherwise LGTM.

File: stamina-core/src/main/scala/stamina/migrations.scala (hunk 5 above)
Comment (psliwa) [reply stamina-core/src/main/scala/stamina/migrations.scala:51]: There is a simple code from tests:

```
val migrator = from[String, V1]
   .to[V2](_ + "V2")
   .backTo[V1](_.replace("V2", ""))
   .to[V3](_ + "V3")
   .backTo[V2](_.replace("V3", ""))
   .to[V4](_ + "V4")
   .backTo[V3](_.replace("V4", ""))
   .to[V5](_ + "V5")
```

The assumption is to define continuous forward migrations, so you cannot define migrations only for version 1 and version 3. You have to define migration also for version 2. So in case when we are going to migrate from version 2 to version 5, we need migrations from 2 to 3, from 3 to 4 and from 4 to 5. Migration from 4 to 5 is under `5` key, it is `toVersion`, so it is needed as well. `canMigrate` function checks in this case existence of `2 -> 3` and `4 -> 5` migrations, because forward migrations are defined continuously. Backward migrations are optional, so they doesn't affect `canMigrate` function at all. If backward migration is not defined for particular version, `identity` function is used.

**Comment by psliwa** [comment]
@raboof ping

**Comment by agemooij** [comment]
@psliwa we are currently waiting on your replies to our comments.

**Comment by psliwa** [comment]
@agemooij Sorry, I am not sure how github conversations work. I added comments long ago, but they were marked as `Pending` and probably you weren't able to see them. Right now you should be able to see them, if you have any more questions - go ahead.

**Comment by agemooij** [comment]
Looks like you started a review but did not finish it. This is a common mistake made by quite a few people I know since Github introduced this feature.

@raboof could you also have a look at the review comments?

**Comment by psliwa** [comment]
@agemooij @raboof ping

**Comment by knirski** [comment]
We will close this PR (@psliwa ping) in favor of https://github.com/scalapenos/stamina/pull/71, which is already proven to be working on production :)
